Doc/tools/extensions/grammar_snippet.py

```diff
@@ -57,51 +57,30 @@ class GrammarSnippetBase(SphinxDirective):
 
     def make_grammar_snippet(
         self, options: dict[str, Any], content: Sequence[str]
-    ) -> list[nodes.paragraph]:
+    ) -> list[addnodes.productionlist]:
         """Create a literal block from options & content."""
 
         group_name = options['group']
-
-        # Docutils elements have a `rawsource` attribute that is supposed to be
-        # set to the original ReST source.
-        # Sphinx does the following with it:
-        # - if it's empty, set it to `self.astext()`
-        # - if it matches `self.astext()` when generating the output,
-        #   apply syntax highlighting (which is based on the plain-text content
-        #   and thus discards internal formatting, like references).
-        # To get around this, we set it to this non-empty string:
-        rawsource = 'You should not see this.'
-
-        literal = nodes.literal_block(
-            rawsource,
-            '',
-            classes=['highlight'],
-        )
-
         node_location = self.get_location()
-        for line in content:
+        productions = [
             self.make_production(
-                line,
-                group_name=group_name,
-                literal=literal,
-                location=node_location,
+                line, group_name=group_name, location=node_location
             )
-        node = nodes.paragraph('', '', literal)
+            for line in content
+        ]
```

Comment:
Is it OK for multi-line productions to be made of multiple `production` nodes?

Reply:
Semantically each grammar production should be contained in one `production` node. I may have misunderstood the new structure though (the only new-style directives are in [top-level components](https://docs.python.org/dev/reference/toplevel_components.html), which are one-line-per-production, no highlighting, etc).

Feel free to push changes to this branch!